### PR TITLE
eth/downloader: retrieve pivot header from local chain if necessary

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR fixes a panic in beacon sync.

**Panic logs**

```
WARN [03-29|23:55:59.938] Local chain is post-merge, waiting for beacon client sync switch-over...
WARN [03-29|23:56:11.942] Ignoring payload with missing parent     number=55128   hash=be29d4..273c2e parent=0d8429..6eba9e
WARN [03-29|23:56:12.044] Ignoring payload with missing parent     number=55129   hash=e6b72a..fe7dc8 parent=be29d4..273c2e
INFO [03-29|23:56:12.176] Left PoW stage
INFO [03-29|23:56:12.176] Forkchoice requested sync to new head    number=55129   hash=e6b72a..fe7dc8
INFO [03-29|23:56:12.322] Looking for peers                        peercount=2 tried=79  static=0
INFO [03-29|23:56:12.366] Forkchoice requested sync to new head    number=55129   hash=e6b72a..fe7dc8
WARN [03-29|23:56:12.366] Beacon chain reorged                     tail=55129 newHead=55129
INFO [03-29|23:56:12.852] Forkchoice requested sync to new head    number=55129   hash=e6b72a..fe7dc8
WARN [03-29|23:56:12.852] Beacon chain reorged                     tail=55129 newHead=55129
INFO [03-29|23:56:13.335] Forkchoice requested sync to new head    number=55129   hash=e6b72a..fe7dc8
WARN [03-29|23:56:13.335] Beacon chain reorged                     tail=55129 newHead=55129
WARN [03-29|23:56:19.918] Header request timed out, dropping peer  peer=f3a3e8ab elapsed=6.574627812s
INFO [03-29|23:56:20.419] Syncing beacon headers                   downloaded=10 left=0           eta=0s
INFO [03-29|23:56:20.420] Disabling direct-ancient mode            origin=18,446,744,073,709,551,615 ancient=55118
WARN [03-29|23:56:20.420] Rewinding blockchain                     target=18,446,744,073,709,551,615
INFO [03-29|23:56:20.423] Loaded most recent local header          number=55127   hash=0d8429..6eba9e td=20,001,479,054,577 age=2w41m28s
INFO [03-29|23:56:20.423] Loaded most recent local full block      number=0       hash=51c7fe..4dd4f8 td=1                  age=53y15h56m
INFO [03-29|23:56:20.423] Loaded most recent local fast block      number=55118   hash=a09656..2c6f47 td=19,986,021,607,809 age=2w42m5s
INFO [03-29|23:56:20.423] Loaded last fast-sync pivot marker       number=0
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x1c0 pc=0x100c24d50]

goroutine 136289 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).processHeaders(0x1400016e180, 0x0, 0x0, 0x0, 0x1)
	/Users/mac/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:1346 +0x810
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func7()
	/Users/mac/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:587 +0x4c
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0x1400016e180, 0x14005e34b40, 0x14013315560)
	/Users/mac/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:608 +0x58
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
	/Users/mac/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:608 +0x98
```

**What's the root cause**

In snap sync, the pivot point is assigned with a remote advertised head minus the safety threshold(64). In beacon sync it obeys the same rules by retrieving the header from the skeleton space. However, if the pivot header is not existent in the skeleton space(e.g. beacon sync only fetch 10 headers as skeleton), the pivot header is left as nil and later set to header(0).

### CODE REVIEW

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -477,15 +480,22 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
...
 		}
 	}
 	// If no pivot block was returned, the head is below the min full block
 	// threshold (i.e. new chain). In that case we won't really snap sync
 	// anyway, but still need a valid pivot block to avoid some code hitting
 	// nil panics on access.
+	var noSnap bool
 	if mode == SnapSync && pivot == nil {
-		pivot = d.blockchain.CurrentBlock().Header()
+		pivot, noSnap = d.blockchain.CurrentBlock().Header(), true
```
Comment (holiman) [inline eth/downloader/downloader.go:498]: What is the semantics of `noSnap`? It is by default `false`, but set to `true` if `mode==SnapSync`, which feels very counter-intuitive to me

Comment (rjl493456442) [reply eth/downloader/downloader.go:498]: It's a special case handled here, where the pivot is nil. In this case it means the sync target is below 64, we don't expect any state sync at all.

And this flag is used in Line 532. In line 532, we try to cap the origin with pivot header to ensure the pivot block is synced from the network. However if we don't expect any state sync, this capping is meaningless.

Maybe I should rename this flag?

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -477,15 +480,22 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
 			return err
 		}
 		if latest.Number.Uint64() > uint64(fsMinFullBlocks) {
-			pivot = d.skeleton.Header(latest.Number.Uint64() - uint64(fsMinFullBlocks))
+			number := latest.Number.Uint64() - uint64(fsMinFullBlocks)
+			pivot = d.skeleton.Header(number)
+			if pivot == nil {
```
Comment (karalabe) [inline eth/downloader/downloader.go:485]: Nit?

```go
if pivot = d.skeleton.Header(number); pivot == nil {
   ...
}
```

Comment (rjl493456442) [reply eth/downloader/downloader.go]: fixed

File: eth/downloader/downloader.go [hunk 3]
```diff
@@ -477,15 +480,22 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
...
 		}
 	}
 	// If no pivot block was returned, the head is below the min full block
 	// threshold (i.e. new chain). In that case we won't really snap sync
 	// anyway, but still need a valid pivot block to avoid some code hitting
 	// nil panics on access.
+	var syncState = true // means whether state sync is required in this cycle
 	if mode == SnapSync && pivot == nil {
-		pivot = d.blockchain.CurrentBlock().Header()
+		pivot, syncState = d.blockchain.CurrentBlock().Header(), false
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply eth/downloader/downloader.go]: Good point. I updated it and also catch another issue.

**Review by karalabe — APPROVED** [review]
SGTM

**Pull request merged**